ls1intum/Artemis
# File Upload Exercise Assessment

### Checklist

### Motivation and Context


This is the last third PR related to adding new exercise type, File Upload Exercise, into Artemis. This PR focuses on allowing tutor to assess student submissions and for students to see the results of assessments.

### Description

Tutors are able to add feedback to submisisons. Tutors can add any number of feedback items, each with description and score. In this case `referencedFeedback` doesn't have any real reference. Also tutors can give general feedback for the submission. Additionally, tutor should not be able to get any data about the student
Students can check the feedback. Each feedback item will be shown separetely to the user. Additionally, user should not be able get any data about the assessor.
Students can also complain/request more feedback in the same way as in other exercises, tutors should be able to see the request and evaluate it.

**NOTE**: I moved some of the code into generic classess, but there is still code duplication in server and client side code. Reducing code duplication will require to change modeling and text exercises and test their functionality. This is out of scope of this PR and will be tackled in a separate PR.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. For File Upload Exercise create a new submission
4. Go to Tutor Exercise Dashboard
5. Start assessment of the File Upload Exercise
6. Submit assessment
7. Check assessment as a student
8. Complain about assessment
9. Evaluate complaint as a tutor
10. Check complaint response as a student

### Screenshots


#### File Upload Exercise in Tutor Exercise Dashboard 
![Screenshot_2019-09-23 Tutor Exercise Dashboard](https://user-images.githubusercontent.com/17262376/65433925-391b8200-de1e-11e9-8708-31acad4ddb58.png)
---
#### Assessment
![Screenshot_2019-09-24 File Upload Exercises](https://user-images.githubusercontent.com/17262376/65552219-2b502480-df24-11e9-8e20-dda1c111b8e9.png)
---
#### Student checks assessment
![Screenshot_2019-09-24 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/65553450-2fca0c80-df27-11e9-8e64-4ddecd88afaa.png)
---
#### Tutor evaluated complaint
![Screenshot_2019-09-25 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/65557038-23977c80-df32-11e9-9a0c-9232968c434f.png)
---
#### Student complaint was accepted
![Screenshot_2019-09-25 File Upload Exercises](https://user-images.githubusercontent.com/17262376/65555963-4e7fd180-df2e-11e9-91ce-b94153697803.png)
